ethstorage/archiver/spec.md

```diff
@@ -0,0 +1,118 @@
+```
+BeaconRootHash->{
+		"data": [
+			{
+			"kv_index": 0,
+			"kzg_commitment": "0x...",
+			"kzg_proof": "0x...",
+			},
+			...
+			]
+		}
+```
+where `BeaconRootHash` can be queried from `/eth/v1/beacon/blocks/{block_id}/root`
+
+2. Sync:
+If the blob retention window has expired and the beacon API is no longer available, the metadata should be synced from peers. In p2p, the `BlobPayload` object will include `BeaconRootHash`, `kzg_commitment`, and `kzg_proof` that can be synced and stored to the key-value table mentioned above.
+
+3. Retrieve:
+It is convenient to load `kzg_commitment` and `kzg_proof` from the key-value . 
+To load blob data, we can we need to compute `dataHash` from `kzg_commitment` and pass `kvIdx` and `dataHash`(as `commit`)  to es-node storage manager:
+ ```
+ func (s *StorageManager) TryRead(kvIdx  uint64, readLen  int, commit  common.Hash) ([]byte, bool, error)
+ ```
+
+ ###  Considerations for Solution 1
+- There will be a flag to control the switch of the archive API, but this flag only controls the external reading interface,  while the downloading and syncing of archive data will always be on.
+- Blobs in the latest block are not supported to be downloaded immediately. You must wait for 2 epochs and L1 to download after the block is finalized.
+
+## Solution 2: on the fly
+
+1. From slot to EL block number:
+Consider `block_id` as the beacon chain slot number or block root, we can obtain the execution layer block number by querying `/eth/v2/beacon/blocks/{block_id}` and accessing `data.message.body.execution_payload.block_number`.
+
+2. Get KV index:
+Next, we can filter logs of the `PutBlob` event by the `block_number` to get `kvIdx` and `dataHash` of the blob uploaded to EthStorage contract:
+```
+event PutBlob(uint256 indexed kvIdx, uint256 indexed kvSize, bytes32 indexed dataHash)
+```
+ 3. Load blob data:
+ Then by passing `kvIdx` and `dataHash`(as `commit`)  we can load blob content using es-node storage manager:
+ ```
+ func (s *StorageManager) TryRead(kvIdx  uint64, readLen  int, commit  common.Hash) ([]byte, bool, error)
+ ```
+ 4. Compute metadata:
+Now that we have blob data, we can generate both `kzg_commitment` and `kzg_proof` by referencing the implementation of the following method:
+```
+func (p *KZGProver) GenerateKZGProof(data []byte, sampleIdx  uint64) ([]byte, error)
+```
+5. Retrieve:
+When the API is queried, the content of blobs is loaded to generate `kzg_commitment` and `kzg_proof` on demand and served by the server.
+
+### Considerations for  Solution 2:
+- It is assumed that the beacon API `/eth/v2/beacon/blocks/{block_id}` is always available for any blocks, and the event log data from the execution layer is always available for any blocks.
+- If the performance is proven to be a problem, we may need to add a local cache/prefetch layer to save the metadata.
+- If the blob transaction hashes are required, they can also be retrieved through standard JSON-RPC API.
+ 
```

Comment:
One scenario we should consider is how an es-node will respond if it receives a request for a BLOB that is not stored by EthStorage.

Reply:
A 404 message will be returned according to spec
```
{
  "code": 404,
  "message": "Block not found"
}
```